src/cart/src/cart/crt_context.c

```diff
@@ -770,51 +702,86 @@ crt_req_timeout_hdlr(struct crt_rpc_priv *rpc_priv)
 	}
 }
 
+struct crt_ep_timeout {
+	d_list_t	 ept_list;
+	uint64_t	 ept_ts_now;
+};
+
+/* timeout the RPCs in inflight queue and waitq in the epi. */
+static int
+crt_ctx_epi_timeout(d_list_t *rlink, void *arg)
+{
+	struct crt_rpc_priv	*rpc_priv, *rpc_next;
+	struct crt_ep_timeout	*ept = arg;
+	struct crt_ep_inflight	*epi;
+	int			 rc = 0;
+
+	D_ASSERT(rlink != NULL);
+	D_ASSERT(arg != NULL);
+	epi = epi_link2ptr(rlink);
+
+	D_MUTEX_LOCK(&epi->epi_mutex);
+
+	/* empty queue, nothing to do */
+	if (d_list_empty(&epi->epi_req_waitq) &&
+	    d_list_empty(&epi->epi_req_q))
+		D_GOTO(out, rc = 0);
+
+	/* timeout RPCs in waitq */
+	d_list_for_each_entry_safe(rpc_priv, rpc_next, &epi->epi_req_waitq,
+				   crp_epi_link) {
+		if (rpc_priv->crp_timeout_ts < ept->ept_ts_now) {
+			/* remove from waitq queue */
+			d_list_del_init(&rpc_priv->crp_epi_link);
+			d_list_add_tail(&rpc_priv->crp_tmp_link,
+					&ept->ept_list);
+		}
+	}
+
+	/* timeout RPCs in inflight queue */
+	d_list_for_each_entry_safe(rpc_priv, rpc_next, &epi->epi_req_q,
+				   crp_epi_link) {
+		if (rpc_priv->crp_timeout_ts < ept->ept_ts_now) {
+			/* remove from inflight queue */
+			d_list_del_init(&rpc_priv->crp_epi_link);
+			d_list_add_tail(&rpc_priv->crp_tmp_link,
+					&ept->ept_list);
+		}
+	}
+
+out:
+	D_MUTEX_UNLOCK(&epi->epi_mutex);
+	return rc;
+}
+
 static void
 crt_context_timeout_check(struct crt_context *crt_ctx)
 {
-	struct crt_rpc_priv		*rpc_priv;
-	struct d_binheap_node		*bh_node;
-	d_list_t			 timeout_list;
-	uint64_t			 ts_now;
+	struct crt_ep_timeout	 ept;
+	struct crt_rpc_priv	*rpc_priv;
 
 	D_ASSERT(crt_ctx != NULL);
 
-	D_INIT_LIST_HEAD(&timeout_list);
-	ts_now = d_timeus_secdiff(0);
+	D_INIT_LIST_HEAD(&ept.ept_list);
+	ept.ept_ts_now = d_timeus_secdiff(0);
 
 	D_MUTEX_LOCK(&crt_ctx->cc_mutex);
-	while (1) {
-		bh_node = d_binheap_root(&crt_ctx->cc_bh_timeout);
-		if (bh_node == NULL)
-			break;
-		rpc_priv = container_of(bh_node, struct crt_rpc_priv,
-					crp_timeout_bp_node);
-		if (rpc_priv->crp_timeout_ts > ts_now)
-			break;
-
-		/* +1 to prevent it from being released in timeout_untrack */
-		RPC_ADDREF(rpc_priv);
-		crt_req_timeout_untrack(rpc_priv);
+	d_hash_table_traverse(&crt_ctx->cc_epi_table, crt_ctx_epi_timeout,
+			      &ept);
+	D_MUTEX_UNLOCK(&crt_ctx->cc_mutex);
 
-		d_list_add_tail(&rpc_priv->crp_tmp_link, &timeout_list);
+	/* handle the timeout RPCs */
+	while ((rpc_priv = d_list_pop_entry(&ept.ept_list, struct crt_rpc_priv,
```

Comment:
It was removed from all queues under lock. So, it will be not accessible for any other processing.

Ok. I will re-think about this additional reference. I don't understand why this need. For now I can not imaginate situation when this can happens. Do you have any example?

Reply:
> Ok. I will re-think about this additional reference. I don't understand why this need. For now I can not imaginate situation when this can happens. Do you have any example?

One scenario would be - server is delaying a response for some reason. A client detects a timeout and removes rpcs from its inflight queue, and as it does that, server responds triggering a completion callback/rpc completion. If there are no extra references held on rpc_priv at that point, the rpc will get free-ed upon completion callback executing 

More generically, since ept_list is referring to rpcs, the rpcs that it contains should have appropriate reference count to account for such reference.